server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/EmailBodyPart.scala

```diff
@@ -19,11 +19,102 @@
-  def serialize(): String = value.toString
+  def of(messageId: MessageId, partId: PartId, entity: Entity): Try[(EmailBodyPart, PartId)] =
+    entity.getBody match {
+      case multipart: Multipart =>
+        val scanResults: Try[List[(Option[EmailBodyPart], PartId)]] = multipart.getBodyParts
+          .asScala.toList
+          .scanLeft[Try[(Option[EmailBodyPart], PartId)]](Success((None, partId)))(traverse(messageId))
+          .sequence
+        val highestPartIdValidation: Try[PartId] = scanResults.map(list => list.map(_._2).reverse.headOption.getOrElse(partId))
+        val childrenValidation: Try[List[EmailBodyPart]] = scanResults.map(list => list.flatMap(_._1))
+
+        zip(childrenValidation, highestPartIdValidation)
+            .flatMap({
+              case (children, highestPartId) => of(messageId, partId, entity, Some(children))
+                .map(part => (part, highestPartId))
+            })
+      case _ => of(messageId, partId, entity, None).map(part => (part, partId))
+    }
+
+  def traverse(messageId: MessageId)(acc: Try[(Option[EmailBodyPart], PartId)], entity: Entity): Try[(Option[EmailBodyPart], PartId)] = {
+    acc.flatMap({
+      case (_, previousPartId) =>
+        val partId = previousPartId.next
+
+        of(messageId, partId, entity)
+          .map({
+            case (part, partId) => (Some(part), partId)
+          })
+    })
+  }
+
+  private def of(messageId: MessageId,
+                 partId: PartId,
+                 entity: Entity,
+                 subParts: Option[List[EmailBodyPart]]): Try[EmailBodyPart] =
+    zip(BlobId.of(messageId, partId), size(entity))
+      .map({
+        case (blobId, size) => EmailBodyPart(
+          partId = partId,
+          blobId = blobId,
+          headers = entity.getHeader.getFields.asScala.toList.map(EmailHeader(_)),
+          size = size,
+          name = Option(entity.getFilename),
+          `type` = entity.getMimeType,
+          charset = Option(entity.getCharset),
+          disposition = Option(entity.getDispositionType),
+          cid = entity.getHeader.getFields("Content-Id").asScala.headOption.map(field => field.getBody),
+          language = entity.getHeader.getFields("Content-Language").asScala.headOption.map(field => field.getBody),
+          location = entity.getHeader.getFields("Content-Location").asScala.headOption.map(field => field.getBody),
+          subParts = subParts)
+      })
+
+  def size(entity: Entity): Try[Size] = {
+    val countingOutputStream: CountingOutputStream = new CountingOutputStream(OutputStream.nullOutputStream())
+    val writer = new DefaultMessageWriter
+    writer.writeEntity(entity, countingOutputStream)
+    refineV[NonNegative](countingOutputStream.getCount) match {
+      case scala.Right(size) => Success(size)
+      case Left(e) => Failure(new IllegalArgumentException(e))
+    }
+  }
+
+  def zip[A, B](a: Try[A], b: Try[B]): Try[(A, B)] = a.flatMap(aValue => b.map(bValue => (aValue, bValue)))
```

Comment:
i think using for comprehension is simpler.
```
  def zip[A, B](a: Try[A], b: Try[B]): Try[(A, B)] = for {
    aValue <- a
    bValue <- b
  } yield (aValue, bValue)
```

Reply:
Nice idea, I did not think about it!